moveit_core/trajectory_processing/src/cartesian_speed.cpp

```diff
@@ -0,0 +1,121 @@
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Authors: Benjamin Scholz, Thies Oelerich */
+
+#include <moveit/trajectory_processing/iterative_time_parameterization.h>
+#include <moveit/trajectory_processing/cartesian_speed.h>
+
+// Name of logger
+const char* LOGGER_NAME = "trajectory_processing.cartesian_speed";
+
+namespace trajectory_processing
+{
+bool limitMaxCartesianLinkSpeed(robot_trajectory::RobotTrajectory& trajectory, const double max_speed,
+                                std::string link_name)
+{
+  // In case the link name is not set, retrieve an end effector name from the
+  // joint model group specified in the robot trajectory
+  if (link_name.empty())
+  {
+    std::vector<std::string> tips;
+    trajectory.getGroup()->getEndEffectorTips(tips);
+    if (tips.empty())
+    {
+      ROS_ERROR_STREAM_NAMED(LOGGER_NAME, "No end effector defined for group attached to trajectory, cannot set max "
+                                          "cartesian link speed without argument.");
+      return false;
+    }
+    link_name = tips[0];
+    if (tips.size() > 1)
+      ROS_INFO_STREAM_NAMED(LOGGER_NAME,
+                            "More than one end effector found, using first one which is '" << link_name << "'");
+  }
+
+  // Get link model based on given end link
+  if (!trajectory.getGroup()->hasLinkModel(link_name))
+    ROS_ERROR_STREAM_NAMED(LOGGER_NAME, "Link model was not specified in the robot trajectory");
+  const moveit::core::LinkModel* link_model = trajectory.getGroup()->getLinkModel(link_name);
+
+  // Call function for speed setting using the created link model
+  return limitMaxCartesianLinkSpeed(trajectory, max_speed, link_model);
+}
+
+bool limitMaxCartesianLinkSpeed(robot_trajectory::RobotTrajectory& trajectory, const double max_speed,
+                                const moveit::core::LinkModel* link_model)
+{
+  if (max_speed <= 0.0)
+  {
+    ROS_ERROR_STREAM_NAMED(LOGGER_NAME, "Link speed must be greater than 0.");
+    return false;
+  }
+
+  size_t num_waypoints = trajectory.getWayPointCount();
+  if (num_waypoints == 0)
+    return false;
+
+  robot_state::RobotStatePtr kinematic_state = trajectory.getFirstWayPointPtr();
+
+  // do forward kinematics to get cartesian positions of link for current waypoint
+  double euclidean_distance, new_time_diff, old_time_diff;
+  std::vector<double> time_diff(num_waypoints - 1, 0.0);
+
+  for (size_t i = 0; i < num_waypoints - 1; i++)
```

Comment:
do we want interpolation or a warning in case joint distance is (very) large? Contrived example: going j1=-180 to j1=180 will have an euclidean distance of 0 but result in fast motion. Actually that does not seem that contrived when I think about ik jumps in cartesian motions